### PR TITLE
Proposal PR: Add DelicateKotlinPoetApi

### PR DESCRIPTION
**Pull request by ZacSweers**
Proposal implementation for https://github.com/square/kotlinpoet/discussions/1087

### CODE REVIEW

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/DelicateKotlinPoetApi.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.squareup.kotlinpoet
+
+/**
+ * Marks declarations in the KotlinPoet that are **delicate** &mdash;
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/DelicateKotlinPoetApi.kt:4]:
```suggestion
 * Marks declarations in the KotlinPoet API that are **delicate** &mdash;
```

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/AnnotationSpec.kt [hunk 2]
```diff
@@ -181,7 +180,13 @@ public class AnnotationSpec private constructor(
   }
 
   public companion object {
-    @JvmStatic @JvmOverloads public fun get(
+    @DelicateKotlinPoetApi(
+      message = "Java reflection APIs don't give complete information on Kotlin types. Consider " +
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/AnnotationSpec.kt:184]: Nice!

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/DelicateKotlinPoetApi.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * Marks declarations in the KotlinPoet API that are **delicate** &mdash;
+ * they have limited use-case and shall be used with care in general code.
+ * Any use of a delicate declaration has to be carefully reviewed to make sure it is
+ * properly used and does not create problems like lossy Java -> Kotlin type parsing.
+ * Carefully read documentation and [message] of any declaration marked as `DelicateKotlinPoetApi`.
+ */
+@MustBeDocumented
+@Retention(value = AnnotationRetention.BINARY)
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/DelicateKotlinPoetApi.kt:11]: Do we need it to be binary?

**Review by Egorand — APPROVED** [review]
Makes sense to me, awesome!

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/DelicateKotlinPoetApi.kt (hunk 3 above)
Comment (ZacSweers) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/DelicateKotlinPoetApi.kt:11]: I believe so, otherwise it wouldn't be visible to consumers right?

Comment (ZacSweers) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/DelicateKotlinPoetApi.kt:11]: Merging as-is but can followup if needed!

Comment (Egorand) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/DelicateKotlinPoetApi.kt:11]: IIRC you only need an annotation in the bytecode if it's supposed to be queried at runtime through reflection, which isn't the case here, the annotation is only processed by the compiler

Comment (ZacSweers) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/DelicateKotlinPoetApi.kt:11]: I think you're thinking of RUNTIME retention for that case. BINARY still ensures it's in bytecode, it just can't be queried at runtime. SOURCE is like it sounds - only in source and omitted everywhere else

Comment (Egorand) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/DelicateKotlinPoetApi.kt:11]: True, sorry!

**Pull request merged**